--- M2Crypto/SSL/__init__.py.old	2011-01-15 12:10:05.000000000 -0700
+++ M2Crypto/SSL/__init__.py	2011-10-18 08:06:44.631797996 -0600
@@ -8,6 +8,9 @@
 class SSLError(Exception): pass
 m2.ssl_init(SSLError)
 
+import Connection
+_Connection = Connection
+
 # M2Crypto.SSL
 from Cipher import Cipher, Cipher_Stack
 from Context import Context
--- M2Crypto/SSL/Connection.py.old	2011-01-15 12:10:05.000000000 -0700
+++ M2Crypto/SSL/Connection.py	2011-10-18 08:12:57.301797993 -0600
@@ -28,6 +28,10 @@
 def _serverPostConnectionCheck(*args, **kw):
     return 1
 
+def log_accept(sock, addr):
+    pass
+
+
 class Connection:
 
     """An SSL connection."""
@@ -154,11 +158,17 @@
         ssl is a new SSL connection object and addr is the address bound to
         the other end of the SSL connection."""
         sock, addr = self.socket.accept()
+        log_accept(sock, addr)
         ssl = Connection(self.ctx, sock)
         ssl.addr = addr
         ssl.setup_ssl()
         ssl.set_accept_state()
-        ssl.accept_ssl()
+        try:
+            ssl.accept_ssl()
+        except Exception, e:
+            e.sock = sock
+            e.addr = addr
+            raise
         check = getattr(self, 'postConnectionCheck', self.serverPostConnectionCheck)
         if check is not None:
             if not check(ssl.get_peer_cert(), ssl.addr[0]):
